src/haywire/http_request.c

```diff
@@ -25,76 +25,102 @@ static const char response_404[] =
   "404 Not Found" CRLF
   ;
 
+static kh_inline khint_t hw_string_hash_func(hw_string* s)
+{
+    khint_t h = s->length > 0 ? (khint_t)*s->value : 0;
+    if (h) for (int i = 0; i < s->length; i++) h = (h << 5) - h + (khint_t)*(s->value + i);
+    return h;
+}
+
+#define hw_string_hash_equal(a, b) (hw_strcmp(a, b) == 0)
+
+KHASH_INIT(hw_string_hashmap, hw_string*,  hw_string*, 1, hw_string_hash_func, hw_string_hash_equal)
 KHASH_MAP_INIT_STR(string_hashmap, char*)
 
 void hw_print_request_headers(http_request* request)
 {
-    const char* k;
-    const char* v;
+    hw_string* k;
+    hw_string* v;
 
-    khash_t(string_hashmap) *h = request->headers;
-    kh_foreach(h, k, v, { printf("KEY: %s VALUE: %s\n", k, v); });
+    khash_t(hw_string_hashmap) *h = request->headers;
+    kh_foreach(h, k, v, {
+        char* key = strndup(k->value, k->length + 1);
+        char* value = strndup(v->value, v->length + 1);
+        printf("KEY: %s VALUE: %s\n", key, value);
+        free(key);
+        free(value);
+    });
 }
 
-void set_header(http_request* request, char* name, char* value)
+void hw_print_body(http_request* request)
 {
-    int ret;
-    khiter_t k;
-    khash_t(string_hashmap) *h = request->headers;
-    k = kh_put(string_hashmap, h, dupstr(name), &ret);
-    kh_value(h, k) = dupstr(value);
+    char* body = strndup(request->body->value, request->body->length + 1);
+    printf("BODY: %s\n", body);
+    free(body);
 }
 
-void* get_header(http_request* request, char* name)
+void* get_header(http_request* request, hw_string* name)
 {
-    khash_t(string_hashmap) *h = request->headers;
-    khiter_t k = kh_get(string_hashmap, h, name);
-    void* val = kh_value(h, k);
+    khash_t(hw_string_hashmap) *h = request->headers;
+    khiter_t k = kh_get(hw_string_hashmap, h, name);
+
+    void* val;
+    
     int is_missing = (k == kh_end(h));
-    if (is_missing)
-    {
+    if (is_missing) {
         val = NULL;
+    } else {
+        val = kh_value(h, k);
     }
     return val;
 }
 
+void set_header(http_request* request, hw_string* name, hw_string* value)
+{
+    int ret;
+    khiter_t k;
+    khash_t(hw_string_hashmap) *h = request->headers;
+    
+    void* prev = get_header(request, name);
+    if (prev) {
+        free(prev);
```

Comment:
This is so someone doesn't bring the server down by sending thousands requests with the same header repeated multiple times. Otherwise, we would leak the previous `hw_string` value.